contrib/pam_zfs_key/pam_zfs_key.c

```diff
@@ -443,9 +475,56 @@ zfs_key_config_free(zfs_key_config_t *config)
 	free(config->homes_prefix);
 }
 
+static int
+find_dsname_by_prop_value(zfs_handle_t *zhp, void *data)
+{
+	zfs_key_config_t *target = data;
+	zfs_type_t type = zfs_get_type(zhp);
+	char mountpoint[ZFS_MAXPROPLEN];
+	char *dsname = NULL;
+
+	// Iterate over any nested datasets.
+	if (zfs_iter_filesystems(zhp, find_dsname_by_prop_value, data) != 0) {
+		zfs_close(zhp);
+		return (1);
+	}
+
+	// Skip any datasets whose type does not match.
+	if ((type & ZFS_TYPE_FILESYSTEM) == 0) {
+		zfs_close(zhp);
+		return (0);
+	}
+
+	dsname = strdup(zfs_get_name(zhp));
+	if (target->homes_prefix != NULL) {
```

Comment:
It looks to me like `target->homes_prefix` can never be NULL.   I'd feel better about either asserting this, or returning immediately with an error if if it's ever NULL.

Reply:
I wonder if `target->homes_prefix` has been freed elsewhere , in that case `strncmp` will probably crash.